File: app/src/main/java/de/xikolo/controllers/main/CertificateListAdapter.java
package de.xikolo.controllers.main;

import android.support.v7.widget.RecyclerView;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.Button;
import android.widget.ImageView;
import android.widget.RelativeLayout;
import android.widget.TextView;

import java.util.ArrayList;
import java.util.List;

import butterknife.BindView;
import butterknife.ButterKnife;
import de.xikolo.App;
import de.xikolo.R;
import de.xikolo.config.GlideApp;
import de.xikolo.models.Course;

public class CertificateListAdapter extends RecyclerView.Adapter<CertificateListAdapter.CertificateViewHolder> {

    public static final String TAG = CertificateListAdapter.class.getSimpleName();

    private List<Course> courseList = new ArrayList<>();
    private OnCertificateCardClickListener callback;

    public CertificateListAdapter(OnCertificateCardClickListener callback) {
        this.callback = callback;
    }

    public void update(List<Course> courseList) {
        this.courseList = courseList;
        notifyDataSetChanged();
    }

    public void clear() {
        courseList.clear();
        this.notifyDataSetChanged();
    }

    @Override
    public int getItemCount() {
        return courseList.size();
    }

    @Override
    public CertificateViewHolder onCreateViewHolder(ViewGroup parent, int viewType) {
        View view = LayoutInflater.from(parent.getContext()).inflate(R.layout.item_certificate_list, parent, false);
        return new CertificateViewHolder(view);
    }

    @Override
    public void onBindViewHolder(CertificateViewHolder holder, int position) {
        final Course course = courseList.get(position);

        holder.textTitle.setText(course.title);

        holder.header.setOnClickListener(v -> callback.onCourseClicked(course.id));

        if (course.imageUrl != null)
            GlideApp.with(App.getInstance()).load(course.imageUrl).into(holder.courseImage);
        else
            holder.courseImage.setVisibility(View.GONE);

        if (!course.certificates.qualifiedCertificateAvailable)
            holder.qualifiedCertificate.setVisibility(View.GONE);
        else
            holder.qualifiedCertificate.setVisibility(View.VISIBLE);

        if (course.certificates.qualifiedCertificateUrl != null) {
            holder.qualifiedCertificateButton.setEnabled(true);
            holder.qualifiedCertificateButton.setOnClickListener(v -> callback.onViewCertificateClicked(course.certificates.qualifiedCertificateUrl));
        } else
            holder.qualifiedCertificateButton.setEnabled(false);


        if (!course.certificates.recordOfAchievementAvailable)
            holder.recordOfAchievement.setVisibility(View.GONE);
        else
            holder.recordOfAchievement.setVisibility(View.VISIBLE);

        if (course.certificates.recordOfAchievementUrl != null) {
            holder.recordOfAchievementButton.setEnabled(true);
            holder.recordOfAchievementButton.setOnClickListener(v -> callback.onViewCertificateClicked(course.certificates.recordOfAchievementUrl));
        } else
            holder.recordOfAchievementButton.setEnabled(false);


        if (!course.certificates.confirmationOfParticipationAvailable)
            holder.confirmationOfParticipation.setVisibility(View.GONE);
        else
            holder.confirmationOfParticipation.setVisibility(View.VISIBLE);

        if (course.certificates.confirmationOfParticipationUrl != null) {
            holder.confirmationOfParticipationButton.setEnabled(true);
            holder.confirmationOfParticipationButton.setOnClickListener(v -> callback.onViewCertificateClicked(course.certificates.confirmationOfParticipationUrl));
        } else
            holder.confirmationOfParticipationButton.setEnabled(false);
    }

    public interface OnCertificateCardClickListener {

        void onViewCertificateClicked(String url);

        void onCourseClicked(String courseId);
    }

    static class CertificateViewHolder extends RecyclerView.ViewHolder {

        @BindView(R.id.certificateHeader) ViewGroup header;
        @BindView(R.id.courseImage) ImageView courseImage;
        @BindView(R.id.textTitle) TextView textTitle;
        @BindView(R.id.qualifiedCertificateContainer) RelativeLayout qualifiedCertificate;
        @BindView(R.id.qualifiedCertificateButton) Button qualifiedCertificateButton;
        @BindView(R.id.recordOfAchievementContainer) RelativeLayout recordOfAchievement;
        @BindView(R.id.recordOfAchievementButton) Button recordOfAchievementButton;
        @BindView(R.id.confirmationOfParticipationContainer) RelativeLayout confirmationOfParticipation;
        @BindView(R.id.confirmationOfParticipationButton) Button confirmationOfParticipationButton;

        public CertificateViewHolder(View view) {
            super(view);
            ButterKnife.bind(this, view);
        }
    }
}

File: app/src/main/java/de/xikolo/controllers/main/CertificateListAdapter.kt
package de.xikolo.controllers.main

import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.LinearLayout
import android.widget.TextView
import androidx.recyclerview.widget.RecyclerView
import butterknife.BindView
import butterknife.ButterKnife
import de.xikolo.App
import de.xikolo.R
import de.xikolo.controllers.helper.DownloadViewHelper
import de.xikolo.models.Course
import de.xikolo.models.DownloadAsset
import de.xikolo.models.dao.EnrollmentDao

class CertificateListAdapter(private val fragment: CertificateListFragment, private val callback: OnCertificateCardClickListener) : RecyclerView.Adapter<CertificateListAdapter.CertificateViewHolder>() {

    companion object {
        val TAG: String = CertificateListAdapter::class.java.simpleName
    }

    private val courseList: MutableList<Course> = mutableListOf()

    fun update(courseList: List<Course>) {
        this.courseList.clear()
        this.courseList.addAll(courseList)
        notifyDataSetChanged()
    }

    override fun getItemCount(): Int {
        return courseList.size
    }

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): CertificateViewHolder {
        val view = LayoutInflater.from(parent.context).inflate(R.layout.item_certificate_list, parent, false)
        return CertificateViewHolder(view)
    }

    override fun onBindViewHolder(holder: CertificateViewHolder, position: Int) {
        val course = courseList[position]

        holder.textTitle.text = course.title

        holder.textTitle.setOnClickListener { _ -> callback.onCourseClicked(course.id) }

        holder.container.removeAllViews()

        fragment.activity?.let { activity ->
            EnrollmentDao.Unmanaged.findForCourse(course.id)?.let { enrollment ->
                if (course.certificates.confirmationOfParticipation.available) {
                    val dvh = DownloadViewHelper(
                        activity,
                        DownloadAsset.Certificate.ConfirmationOfParticipation(
                            enrollment.certificates.confirmationOfParticipationUrl,
                            course
                        ),
                        App.instance.getString(R.string.course_confirmation_of_participation),
                        null,
                        App.instance.getString(R.string.course_certificate_not_achieved)
                    )

                    holder.container.addView(dvh.view)
                }

                if (course.certificates.recordOfAchievement.available) {
                    val dvh = DownloadViewHelper(
                        activity,
                        DownloadAsset.Certificate.RecordOfAchievement(
                            enrollment.certificates.recordOfAchievementUrl,
                            course
                        ),
                        App.instance.getString(R.string.course_record_of_achievement),
                        null,
                        App.instance.getString(R.string.course_certificate_not_achieved)
                    )

                    holder.container.addView(dvh.view)
                }

                if (course.certificates.qualifiedCertificate.available) {
                    val dvh = DownloadViewHelper(
                        activity,
                        DownloadAsset.Certificate.QualifiedCertificate(
                            enrollment.certificates.qualifiedCertificateUrl,
                            course
                        ),
                        App.instance.getString(R.string.course_qualified_certificate),
                        null,
                        App.instance.getString(R.string.course_certificate_not_achieved)
                    )

                    holder.container.addView(dvh.view)
                }
            }
        }
    }

    interface OnCertificateCardClickListener {
        fun onCourseClicked(courseId: String)
    }

    class CertificateViewHolder(view: View) : RecyclerView.ViewHolder(view) {

        @BindView(R.id.textTitle)
        lateinit var textTitle: TextView

        @BindView(R.id.container)
        lateinit var container: LinearLayout

        init {
            ButterKnife.bind(this, view)
        }

    }

}
